{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Nubiancodingdelight/ACS-Project-Repository-/blob/main/Generating_VIS.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UitA4QM35UrE",
        "outputId": "f3ed452c-fde3-4049-c85e-f76157d8cf40"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "#mounts google drive\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#set path\n",
        "%cd /content/drive/MyDrive/ACS_Research/VISDB_Data/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DgBjRq1H_vYS",
        "outputId": "d44c6ddd-1532-4da8-ed35-b0a7cb6577a9"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/ACS_Research/VISDB_Data\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#import\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import os\n",
        "from pyfaidx import Fasta"
      ],
      "metadata": {
        "id": "tZ5SZAIFAKAv"
      },
      "execution_count": 30,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install Bio pyfaidx"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wFafUZikMydT",
        "outputId": "ac010755-81b4-4944-a48b-8531f14b24a5"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting Bio\n",
            "  Downloading bio-1.7.1-py3-none-any.whl.metadata (5.7 kB)\n",
            "Collecting pyfaidx\n",
            "  Downloading pyfaidx-0.8.1.3-py3-none-any.whl.metadata (25 kB)\n",
            "Collecting biopython>=1.80 (from Bio)\n",
            "  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Collecting gprofiler-official (from Bio)\n",
            "  Downloading gprofiler_official-1.0.0-py3-none-any.whl.metadata (11 kB)\n",
            "Collecting mygene (from Bio)\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl.metadata (10 kB)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (from Bio) (2.2.2)\n",
            "Requirement already satisfied: pooch in /usr/local/lib/python3.11/dist-packages (from Bio) (1.8.2)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from Bio) (2.32.3)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.11/dist-packages (from Bio) (4.67.1)\n",
            "Requirement already satisfied: importlib-metadata in /usr/local/lib/python3.11/dist-packages (from pyfaidx) (8.6.1)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.11/dist-packages (from pyfaidx) (24.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython>=1.80->Bio) (2.0.2)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.11/dist-packages (from importlib-metadata->pyfaidx) (3.21.0)\n",
            "Collecting biothings-client>=0.2.6 (from mygene->Bio)\n",
            "  Downloading biothings_client-0.4.1-py3-none-any.whl.metadata (10 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas->Bio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas->Bio) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas->Bio) (2025.1)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.11/dist-packages (from pooch->Bio) (4.3.7)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->Bio) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->Bio) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->Bio) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->Bio) (2025.1.31)\n",
            "Requirement already satisfied: httpx>=0.22.0 in /usr/local/lib/python3.11/dist-packages (from biothings-client>=0.2.6->mygene->Bio) (0.28.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas->Bio) (1.17.0)\n",
            "Requirement already satisfied: anyio in /usr/local/lib/python3.11/dist-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene->Bio) (4.9.0)\n",
            "Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.11/dist-packages (from httpx>=0.22.0->biothings-client>=0.2.6->mygene->Bio) (1.0.7)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/lib/python3.11/dist-packages (from httpcore==1.*->httpx>=0.22.0->biothings-client>=0.2.6->mygene->Bio) (0.14.0)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio->httpx>=0.22.0->biothings-client>=0.2.6->mygene->Bio) (1.3.1)\n",
            "Requirement already satisfied: typing_extensions>=4.5 in /usr/local/lib/python3.11/dist-packages (from anyio->httpx>=0.22.0->biothings-client>=0.2.6->mygene->Bio) (4.12.2)\n",
            "Downloading bio-1.7.1-py3-none-any.whl (280 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m281.0/281.0 kB\u001b[0m \u001b[31m17.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pyfaidx-0.8.1.3-py3-none-any.whl (28 kB)\n",
            "Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m64.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
            "Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Downloading biothings_client-0.4.1-py3-none-any.whl (46 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.7/46.7 kB\u001b[0m \u001b[31m2.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython, pyfaidx, gprofiler-official, biothings-client, mygene, Bio\n",
            "Successfully installed Bio-1.7.1 biopython-1.85 biothings-client-0.4.1 gprofiler-official-1.0.0 mygene-3.2.2 pyfaidx-0.8.1.3\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Extract sequence from virus"
      ],
      "metadata": {
        "id": "lm38Gg9U-HAS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "def virus_seq_extract(name, ref, start, stop):\n",
        "  path = '/content/drive/MyDrive/ACS_Research/VISDB_Data/Viral_Reference/' +name+ \"/\"+ ref + \".fasta\"\n",
        "  for record in SeqIO.parse(path, \"fasta\"):\n",
        "      extracted_seq = record.seq[start-1:stop]\n",
        "  return extracted_seq"
      ],
      "metadata": {
        "id": "FNDCBtmP-GZL"
      },
      "execution_count": 26,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def virus_seq_extract(name, ref, start, stop):\n",
        "  path = '/content/drive/MyDrive/ACS_Research/VISDB_Data/Viral_Reference/' +name+ \"/\"+ \"sequence.fasta\"\n",
        "  virus_gene = Fasta(path)\n",
        "  extracted_seq = virus_gene[ref][start-1:stop]\n",
        "  return extracted_seq\n",
        " # for record in SeqIO.parse(path, \"fasta\"):\n",
        "  #    extracted_seq = record.seq[start-1:stop]\n",
        "  #return extracted_seq"
      ],
      "metadata": {
        "id": "Yh1y-4gHUIyn"
      },
      "execution_count": 37,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#testing\n",
        "print(virus_seq_extract('HBV', 'X70185.1',50, 100))"
      ],
      "metadata": {
        "id": "4BQ2L8anSixt",
        "outputId": "507c7eaa-950d-40f2-f4c7-f4d34da86695",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "TATCTTCCTGCTGGTGGCTCCAGTTCAGGAACAGTAAACCCTGCTCCGAAT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Extract Sequence from Human Reference 100 bp upstream and 100 bp downstream"
      ],
      "metadata": {
        "id": "sazjWlI--KHH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def human_seq_extract(ref, chr, start, stop):\n",
        "  seq_list = []\n",
        "  #hg19\n",
        "  if ref in ('hg19':\n",
        "    path = '/content/drive/MyDrive/ACS_Research/VISDB_Data/Human_Reference/hg19/uncompressed/' + chr + '.fa'\n",
        "    genes = Fasta(path)\n",
        "    if start < stop:\n",
        "      seq_list.append(genes[0][start-100:start])\n",
        "      seq_list.append(genes[0][stop:stop+100])\n",
        "    else:\n",
        "      seq_list.append(genes[0][start:start+100])\n",
        "      seq_list.append(genes[0][stop-100:stop])\n",
        "  else:\n",
        "    #hg38\n",
        "    path = '/content/drive/MyDrive/ACS_Research/VISDB_Data/Human_Reference/hg38/uncompressed/' + chr + '.fa'\n",
        "    genes = Fasta(path)\n",
        "    if start < stop:\n",
        "      seq_list.append(genes[0][start-100:start])\n",
        "      seq_list.append(genes[0][stop:stop+100])\n",
        "    else:\n",
        "      seq_list.append(genes[0][start:start+100])\n",
        "      seq_list.append(genes[0][stop-100:stop])\n",
        "  return seq_list"
      ],
      "metadata": {
        "id": "dZZAzNYG-R38"
      },
      "execution_count": 31,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#testing\n",
        "print(human_seq_extract('hg19', 'chr1', 100, 200))"
      ],
      "metadata": {
        "id": "OyqDBoCrTN0d",
        "outputId": "ba48450d-372e-4205-9316-97435005423b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[>chr1:1-100\n",
            "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN, >chr1:201-300\n",
            "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Load Data"
      ],
      "metadata": {
        "id": "J6NVK_fAA66R"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#load xslx\n",
        "vis_df = pd.read_excel('Supplementary Table 1.xlsx')\n",
        "vis_df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 434
        },
        "id": "E28mDMEfAFI2",
        "outputId": "f99a21ad-f4e2-406b-9532-1938e067f066"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  Virus Virus subtype Reference virus genome  Begin breakpoint  \\\n",
              "0   HBV           NaN             FJ562281.1            1782.0   \n",
              "1   HBV           NaN             GQ377546.1            1820.0   \n",
              "2   HBV           NaN             KC792890.1            1471.0   \n",
              "3   HBV           NaN             FJ622847.1             689.0   \n",
              "4   HBV           NaN             KF166416.1            1807.0   \n",
              "\n",
              "   Stop breakpoint Chromosome  Locus Reference human genome  Begin location  \\\n",
              "0              NaN       chr9  q31.1                 GRCh38     100940295.0   \n",
              "1              NaN       chr9  q31.1                 GRCh38     101030026.0   \n",
              "2              NaN       chr3  p26.3                 GRCh38       2002381.0   \n",
              "3              NaN       chr2    q35                 GRCh38     216281034.0   \n",
              "4              NaN      chr16  q12.1                 GRCh38      51320015.0   \n",
              "\n",
              "   End location Sample                    Disease                     Method  \\\n",
              "0           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "1           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "2           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "3           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "4           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "\n",
              "   PubMed ID Target gene  \n",
              "0   30535432      CORO2A  \n",
              "1   30535432         NaN  \n",
              "2   30535432         NaN  \n",
              "3   30535432         NaN  \n",
              "4   30535432         NaN  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-077d6fef-3678-412c-877d-2b8874ef3f21\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Virus</th>\n",
              "      <th>Virus subtype</th>\n",
              "      <th>Reference virus genome</th>\n",
              "      <th>Begin breakpoint</th>\n",
              "      <th>Stop breakpoint</th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Locus</th>\n",
              "      <th>Reference human genome</th>\n",
              "      <th>Begin location</th>\n",
              "      <th>End location</th>\n",
              "      <th>Sample</th>\n",
              "      <th>Disease</th>\n",
              "      <th>Method</th>\n",
              "      <th>PubMed ID</th>\n",
              "      <th>Target gene</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>FJ562281.1</td>\n",
              "      <td>1782.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr9</td>\n",
              "      <td>q31.1</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>100940295.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>CORO2A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GQ377546.1</td>\n",
              "      <td>1820.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr9</td>\n",
              "      <td>q31.1</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>101030026.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>KC792890.1</td>\n",
              "      <td>1471.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr3</td>\n",
              "      <td>p26.3</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>2002381.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>FJ622847.1</td>\n",
              "      <td>689.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr2</td>\n",
              "      <td>q35</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>216281034.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>KF166416.1</td>\n",
              "      <td>1807.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr16</td>\n",
              "      <td>q12.1</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>51320015.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-077d6fef-3678-412c-877d-2b8874ef3f21')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-077d6fef-3678-412c-877d-2b8874ef3f21 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-077d6fef-3678-412c-877d-2b8874ef3f21');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-82254c7d-a263-4f51-8ec1-170709134e4c\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-82254c7d-a263-4f51-8ec1-170709134e4c')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-82254c7d-a263-4f51-8ec1-170709134e4c button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "vis_df",
              "summary": "{\n  \"name\": \"vis_df\",\n  \"rows\": 20588,\n  \"fields\": [\n    {\n      \"column\": \"Virus\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"HBV\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Virus subtype\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"B2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Reference virus genome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 40,\n        \"samples\": [\n          \" V01460.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Begin breakpoint\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 210338.7192678196,\n        \"min\": 1.0,\n        \"max\": 18003056.0,\n        \"num_unique_values\": 2693,\n        \"samples\": [\n          3218.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Stop breakpoint\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 654.6863577028345,\n        \"min\": 31.0,\n        \"max\": 3646.0,\n        \"num_unique_values\": 1046,\n        \"samples\": [\n          1206.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Chromosome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 56,\n        \"samples\": [\n          \"chr9\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Locus\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 187,\n        \"samples\": [\n          \"p16\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Reference human genome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 27,\n        \"samples\": [\n          \"Feb_2009 build\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Begin location\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 61091376.55399811,\n        \"min\": 54.0,\n        \"max\": 2000084860.0,\n        \"num_unique_values\": 18445,\n        \"samples\": [\n          54182535.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"End location\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 60794458.37028902,\n        \"min\": 27.0,\n        \"max\": 249241007.0,\n        \"num_unique_values\": 2008,\n        \"samples\": [\n          129392031.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Sample\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 12,\n        \"samples\": [\n          \"Cell line\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Disease\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \" Chronic HBV (CHB) infection\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Method\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 22,\n        \"samples\": [\n          \"PCR and Sanger sequencing\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"PubMed ID\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 2664585,\n        \"min\": 11593432,\n        \"max\": 30535432,\n        \"num_unique_values\": 45,\n        \"samples\": [\n          14712219\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Target gene\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3942,\n        \"samples\": [\n          \"RAI1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Transform Data"
      ],
      "metadata": {
        "id": "lsIU_RuFA-76"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def lower_case_and_underscore(df):\n",
        "  \"\"\"Lowercases all column names and replaces spaces with underscores.\"\"\"\n",
        "  new_columns = [col.lower().replace(' ', '_') for col in df.columns]\n",
        "  df.columns = new_columns\n",
        "  return df"
      ],
      "metadata": {
        "id": "H64rswrjBB5n"
      },
      "execution_count": 39,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "vis_df = lower_case_and_underscore(vis_df)\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 400
        },
        "id": "PY9fYTgGAkfy",
        "outputId": "830e8b41-4621-4088-b403-6da3bbeb8309"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  virus virus_subtype reference_virus_genome  begin_breakpoint  \\\n",
              "0   HBV           NaN             FJ562281.1            1782.0   \n",
              "1   HBV           NaN             GQ377546.1            1820.0   \n",
              "2   HBV           NaN             KC792890.1            1471.0   \n",
              "3   HBV           NaN             FJ622847.1             689.0   \n",
              "4   HBV           NaN             KF166416.1            1807.0   \n",
              "\n",
              "   stop_breakpoint chromosome  locus reference_human_genome  begin_location  \\\n",
              "0              NaN       chr9  q31.1                 GRCh38     100940295.0   \n",
              "1              NaN       chr9  q31.1                 GRCh38     101030026.0   \n",
              "2              NaN       chr3  p26.3                 GRCh38       2002381.0   \n",
              "3              NaN       chr2    q35                 GRCh38     216281034.0   \n",
              "4              NaN      chr16  q12.1                 GRCh38      51320015.0   \n",
              "\n",
              "   end_location sample                    disease                     method  \\\n",
              "0           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "1           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "2           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "3           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "4           NaN  Tumor  Hepatocellular  carcinoma  PCR and Sanger sequencing   \n",
              "\n",
              "   pubmed_id target_gene  \n",
              "0   30535432      CORO2A  \n",
              "1   30535432         NaN  \n",
              "2   30535432         NaN  \n",
              "3   30535432         NaN  \n",
              "4   30535432         NaN  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-965b8f96-b2f1-4954-81d0-878557eccfd7\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>virus</th>\n",
              "      <th>virus_subtype</th>\n",
              "      <th>reference_virus_genome</th>\n",
              "      <th>begin_breakpoint</th>\n",
              "      <th>stop_breakpoint</th>\n",
              "      <th>chromosome</th>\n",
              "      <th>locus</th>\n",
              "      <th>reference_human_genome</th>\n",
              "      <th>begin_location</th>\n",
              "      <th>end_location</th>\n",
              "      <th>sample</th>\n",
              "      <th>disease</th>\n",
              "      <th>method</th>\n",
              "      <th>pubmed_id</th>\n",
              "      <th>target_gene</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>FJ562281.1</td>\n",
              "      <td>1782.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr9</td>\n",
              "      <td>q31.1</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>100940295.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>CORO2A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GQ377546.1</td>\n",
              "      <td>1820.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr9</td>\n",
              "      <td>q31.1</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>101030026.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>KC792890.1</td>\n",
              "      <td>1471.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr3</td>\n",
              "      <td>p26.3</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>2002381.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>FJ622847.1</td>\n",
              "      <td>689.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr2</td>\n",
              "      <td>q35</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>216281034.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>KF166416.1</td>\n",
              "      <td>1807.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>chr16</td>\n",
              "      <td>q12.1</td>\n",
              "      <td>GRCh38</td>\n",
              "      <td>51320015.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>PCR and Sanger sequencing</td>\n",
              "      <td>30535432</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-965b8f96-b2f1-4954-81d0-878557eccfd7')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-965b8f96-b2f1-4954-81d0-878557eccfd7 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-965b8f96-b2f1-4954-81d0-878557eccfd7');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-ea650e53-da28-49e9-8c4d-f09f79e5d34b\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-ea650e53-da28-49e9-8c4d-f09f79e5d34b')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-ea650e53-da28-49e9-8c4d-f09f79e5d34b button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "vis_df",
              "summary": "{\n  \"name\": \"vis_df\",\n  \"rows\": 20588,\n  \"fields\": [\n    {\n      \"column\": \"virus\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"HBV\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"virus_subtype\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"B2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"reference_virus_genome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 40,\n        \"samples\": [\n          \" V01460.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"begin_breakpoint\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 210338.7192678196,\n        \"min\": 1.0,\n        \"max\": 18003056.0,\n        \"num_unique_values\": 2693,\n        \"samples\": [\n          3218.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"stop_breakpoint\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 654.6863577028345,\n        \"min\": 31.0,\n        \"max\": 3646.0,\n        \"num_unique_values\": 1046,\n        \"samples\": [\n          1206.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chromosome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 56,\n        \"samples\": [\n          \"chr9\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"locus\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 187,\n        \"samples\": [\n          \"p16\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"reference_human_genome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 27,\n        \"samples\": [\n          \"Feb_2009 build\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"begin_location\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 61091376.55399811,\n        \"min\": 54.0,\n        \"max\": 2000084860.0,\n        \"num_unique_values\": 18445,\n        \"samples\": [\n          54182535.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"end_location\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 60794458.37028902,\n        \"min\": 27.0,\n        \"max\": 249241007.0,\n        \"num_unique_values\": 2008,\n        \"samples\": [\n          129392031.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"sample\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 12,\n        \"samples\": [\n          \"Cell line\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"disease\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \" Chronic HBV (CHB) infection\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"method\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 22,\n        \"samples\": [\n          \"PCR and Sanger sequencing\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pubmed_id\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 2664585,\n        \"min\": 11593432,\n        \"max\": 30535432,\n        \"num_unique_values\": 45,\n        \"samples\": [\n          14712219\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"target_gene\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3942,\n        \"samples\": [\n          \"RAI1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# prompt: filter out rows where begin_breakpoint or stop breakpoint  and also begin and end location are NA\n",
        "\n",
        "vis_df = vis_df.dropna(subset=['begin_breakpoint', 'stop_breakpoint', 'begin_location', 'end_location', 'reference_virus_genome'], how='any')\n",
        "vis_df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 313
        },
        "id": "nnRB-Q5MBLx0",
        "outputId": "4b081ebd-3774-4ff5-b38b-2bcdd7eea38b"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "    virus virus_subtype reference_virus_genome  begin_breakpoint  \\\n",
              "346   HBV           NaN               X70185.1            1720.0   \n",
              "347   HBV           NaN               X70185.1            1246.0   \n",
              "348   HBV           NaN               X70185.1            1603.0   \n",
              "349   HBV           NaN               X70185.1            1623.0   \n",
              "350   HBV           NaN               X70185.1            1638.0   \n",
              "\n",
              "     stop_breakpoint chromosome  locus reference_human_genome  begin_location  \\\n",
              "346           1603.0       chr1  q21.1            GRCh37/hg19           420.0   \n",
              "347           1390.0       chr1  p21.3            GRCh37/hg19      97732036.0   \n",
              "348           1664.0       chr1  q31.3            GRCh37/hg19     194161891.0   \n",
              "349           1717.0       chr2  q31.1            GRCh37/hg19     171450967.0   \n",
              "350           1717.0       chr2  q31.1            GRCh37/hg19     171450967.0   \n",
              "\n",
              "     end_location sample                    disease      method  pubmed_id  \\\n",
              "346         437.0   cell  Hepatocellular  carcinoma  invPCR+NAH   28414318   \n",
              "347    97732056.0   cell  Hepatocellular  carcinoma  invPCR+NAH   28414318   \n",
              "348   194161910.0   cell  Hepatocellular  carcinoma  invPCR+NAH   28414318   \n",
              "349   171450816.0   cell  Hepatocellular  carcinoma  invPCR+NAH   28414318   \n",
              "350   171450816.0   cell  Hepatocellular  carcinoma  invPCR+NAH   28414318   \n",
              "\n",
              "    target_gene  \n",
              "346      ANP32E  \n",
              "347        DPYD  \n",
              "348         NaN  \n",
              "349       MYO3B  \n",
              "350       MYO3B  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-eedfad7f-2984-4256-b204-b8d782d9bd55\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>virus</th>\n",
              "      <th>virus_subtype</th>\n",
              "      <th>reference_virus_genome</th>\n",
              "      <th>begin_breakpoint</th>\n",
              "      <th>stop_breakpoint</th>\n",
              "      <th>chromosome</th>\n",
              "      <th>locus</th>\n",
              "      <th>reference_human_genome</th>\n",
              "      <th>begin_location</th>\n",
              "      <th>end_location</th>\n",
              "      <th>sample</th>\n",
              "      <th>disease</th>\n",
              "      <th>method</th>\n",
              "      <th>pubmed_id</th>\n",
              "      <th>target_gene</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>346</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>X70185.1</td>\n",
              "      <td>1720.0</td>\n",
              "      <td>1603.0</td>\n",
              "      <td>chr1</td>\n",
              "      <td>q21.1</td>\n",
              "      <td>GRCh37/hg19</td>\n",
              "      <td>420.0</td>\n",
              "      <td>437.0</td>\n",
              "      <td>cell</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>invPCR+NAH</td>\n",
              "      <td>28414318</td>\n",
              "      <td>ANP32E</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>347</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>X70185.1</td>\n",
              "      <td>1246.0</td>\n",
              "      <td>1390.0</td>\n",
              "      <td>chr1</td>\n",
              "      <td>p21.3</td>\n",
              "      <td>GRCh37/hg19</td>\n",
              "      <td>97732036.0</td>\n",
              "      <td>97732056.0</td>\n",
              "      <td>cell</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>invPCR+NAH</td>\n",
              "      <td>28414318</td>\n",
              "      <td>DPYD</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>348</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>X70185.1</td>\n",
              "      <td>1603.0</td>\n",
              "      <td>1664.0</td>\n",
              "      <td>chr1</td>\n",
              "      <td>q31.3</td>\n",
              "      <td>GRCh37/hg19</td>\n",
              "      <td>194161891.0</td>\n",
              "      <td>194161910.0</td>\n",
              "      <td>cell</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>invPCR+NAH</td>\n",
              "      <td>28414318</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>349</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>X70185.1</td>\n",
              "      <td>1623.0</td>\n",
              "      <td>1717.0</td>\n",
              "      <td>chr2</td>\n",
              "      <td>q31.1</td>\n",
              "      <td>GRCh37/hg19</td>\n",
              "      <td>171450967.0</td>\n",
              "      <td>171450816.0</td>\n",
              "      <td>cell</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>invPCR+NAH</td>\n",
              "      <td>28414318</td>\n",
              "      <td>MYO3B</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>350</th>\n",
              "      <td>HBV</td>\n",
              "      <td>NaN</td>\n",
              "      <td>X70185.1</td>\n",
              "      <td>1638.0</td>\n",
              "      <td>1717.0</td>\n",
              "      <td>chr2</td>\n",
              "      <td>q31.1</td>\n",
              "      <td>GRCh37/hg19</td>\n",
              "      <td>171450967.0</td>\n",
              "      <td>171450816.0</td>\n",
              "      <td>cell</td>\n",
              "      <td>Hepatocellular  carcinoma</td>\n",
              "      <td>invPCR+NAH</td>\n",
              "      <td>28414318</td>\n",
              "      <td>MYO3B</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-eedfad7f-2984-4256-b204-b8d782d9bd55')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-eedfad7f-2984-4256-b204-b8d782d9bd55 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-eedfad7f-2984-4256-b204-b8d782d9bd55');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-a39e529a-7b35-44e3-b580-f5eb7775e112\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-a39e529a-7b35-44e3-b580-f5eb7775e112')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-a39e529a-7b35-44e3-b580-f5eb7775e112 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "vis_df",
              "summary": "{\n  \"name\": \"vis_df\",\n  \"rows\": 1509,\n  \"fields\": [\n    {\n      \"column\": \"virus\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"HBV\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"virus_subtype\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"C\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"reference_virus_genome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 11,\n        \"samples\": [\n          \"AY033073.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"begin_breakpoint\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 751.759081783155,\n        \"min\": 1.0,\n        \"max\": 3210.0,\n        \"num_unique_values\": 864,\n        \"samples\": [\n          1470.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"stop_breakpoint\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 726.2967002344485,\n        \"min\": 31.0,\n        \"max\": 3646.0,\n        \"num_unique_values\": 859,\n        \"samples\": [\n          1597.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chromosome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 31,\n        \"samples\": [\n          \"GL000237.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"locus\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 30,\n        \"samples\": [\n          \"p\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"reference_human_genome\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 11,\n        \"samples\": [\n          \"U96773\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"begin_location\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 77493193.55510998,\n        \"min\": 54.0,\n        \"max\": 2000084860.0,\n        \"num_unique_values\": 1457,\n        \"samples\": [\n          36212916.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"end_location\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 59638565.721075185,\n        \"min\": 27.0,\n        \"max\": 249241007.0,\n        \"num_unique_values\": 1475,\n        \"samples\": [\n          139027412.0\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"sample\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"Tumor\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"disease\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"Hepatocellular  carcinoma\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"method\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 9,\n        \"samples\": [\n          \"RT-PCR+Long-range PCR\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"pubmed_id\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 3240855,\n        \"min\": 11593432,\n        \"max\": 30531861,\n        \"num_unique_values\": 12,\n        \"samples\": [\n          12802289\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"target_gene\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 223,\n        \"samples\": [\n          \"TET1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "vis_df['reference_human_genome'].unique()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2ieyZkqGVSsE",
        "outputId": "940228a7-33c4-4393-e678-ca696cb2481b"
      },
      "execution_count": 44,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['GRCh37/hg19', 'GRCh37.p10', 'GRCh37', 'GRCh37.55', 'NT011005',\n",
              "       'U96773', 'AF128893', 'AC009318', 'AL360074', 'AL035461',\n",
              "       ' AY007685'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 44
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print('Number of Rows: ' + str(len(vis_df)))\n",
        "vis_df['reference_virus_genome'].unique()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lrMNJCo3A11U",
        "outputId": "903eb19d-58ad-4803-fe43-647eaa09fd15"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of Rows: 1509\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['X70185.1', 'AB033556.1', 'X75657.1', 'NC_003977.1', 'NC_003977',\n",
              "       'AY033073.1', 'V01460.1', 'X0276', 'V00866', 'AF068756 ',\n",
              "       'AF330110'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Concatenate Viral and Human Sequence and Insert into"
      ],
      "metadata": {
        "id": "YWBjwZNPGHi1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#loop through data\n",
        "seq_list =[]\n",
        "for ind, row in vis_df.iterrows():\n",
        "  virus_name = row['virus']\n",
        "  virus_ref = row['reference_virus_genome']\n",
        "  hum_ref = row['reference_human_genome']\n",
        "  chrom = row['chromosome']\n",
        "  human_seq = human_seq_extract(hum_ref, chrom, row['begin_location'],row['end_location'])\n",
        "  viral_seq = virus_seq_extract(virus, virus_ref, row['begin_breakpoint'], row['stop_breakpoint'])\n",
        "  spliced_seq = human_seq[0] + viral_seq + human_seq[1]\n",
        "  seq_list.append([virus_name, virus_ref, hum_ref, spliced_seq, viral_seq, human_seq[0], human_seq[1]])\n",
        "\n"
      ],
      "metadata": {
        "id": "vUpW4sGwF-eB",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 495
        },
        "outputId": "cf143448-23af-4aeb-e968-eed59876752e"
      },
      "execution_count": 43,
      "outputs": [
        {
          "output_type": "error",
          "ename": "FastaNotFoundError",
          "evalue": "Cannot read FASTA from file /content/drive/MyDrive/ACS_Research/VISDB_Data/Human_Reference/hg38/uncompressed/chr1.fa",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pyfaidx/__init__.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, duplicate_action, filt_function, one_based_attributes, read_long_names, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m    390\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 391\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r+b'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mmutable\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m'rb'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    392\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mIOError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/content/drive/MyDrive/ACS_Research/VISDB_Data/Human_Reference/hg38/uncompressed/chr1.fa'",
            "\nDuring handling of the above exception, another exception occurred:\n",
            "\u001b[0;31mFastaNotFoundError\u001b[0m                        Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-43-327f1e94660c>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m   \u001b[0mhum_ref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'reference_human_genome'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m   \u001b[0mchrom\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chromosome'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m   \u001b[0mhuman_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhuman_seq_extract\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhum_ref\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchrom\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'begin_location'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'end_location'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m   \u001b[0mviral_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvirus_seq_extract\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvirus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvirus_ref\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'begin_breakpoint'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'stop_breakpoint'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m   \u001b[0mspliced_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhuman_seq\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mviral_seq\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mhuman_seq\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-31-e64c701f9d87>\u001b[0m in \u001b[0;36mhuman_seq_extract\u001b[0;34m(ref, chr, start, stop)\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0;31m#hg38\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0mpath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'/content/drive/MyDrive/ACS_Research/VISDB_Data/Human_Reference/hg38/uncompressed/'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mchr\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'.fa'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m     \u001b[0mgenes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFasta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mstop\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m       \u001b[0mseq_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pyfaidx/__init__.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, filt_function, one_based_attributes, read_long_names, duplicate_action, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m   1088\u001b[0m         \"\"\"\n\u001b[1;32m   1089\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmutable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmutable\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1090\u001b[0;31m         self.faidx = Faidx(\n\u001b[0m\u001b[1;32m   1091\u001b[0m             \u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1092\u001b[0m             \u001b[0mindexname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindexname\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pyfaidx/__init__.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, indexname, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, duplicate_action, filt_function, one_based_attributes, read_long_names, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m    391\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r+b'\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mmutable\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0;34m'rb'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    392\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mIOError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 393\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mFastaNotFoundError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Cannot read FASTA from file %s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    394\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    395\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFastaNotFoundError\u001b[0m: Cannot read FASTA from file /content/drive/MyDrive/ACS_Research/VISDB_Data/Human_Reference/hg38/uncompressed/chr1.fa"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#make seq_list a dataframe save as a csv\n",
        "seq_df = pd.DataFrame(seq_list, columns=['virus', 'virus_ref', 'human_ref', 'spliced_seq', 'viral_seq', 'human_seq_upstream', 'human_seq_downstream'])\n"
      ],
      "metadata": {
        "id": "YLzLRB52Ja3Z"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}